cranelift-wasm/src/environ/spec.rs

```diff
@@ -266,6 +266,81 @@ pub trait FuncEnvironment {
         heap: ir::Heap,
     ) -> WasmResult<ir::Value>;
 
+    /// Translate a `memory.copy` WebAssembly instruction.
+    ///
+    /// The `index` provided identifies the linear memory to query, and `heap` is the heap reference
+    /// returned by `make_heap` for the same index.
+    fn translate_memory_copy(
+        &mut self,
+        pos: FuncCursor,
+        index: MemoryIndex,
+        heap: ir::Heap,
+        dst: ir::Value,
+        src: ir::Value,
+        len: ir::Value,
```

Comment:
I wonder if in the interest of having a nice API, we could introduce a structure to pass parameters to these functions, so there's no doubt about the ordering (since Rust doesn't have named parameters). It might not be a big deal, since IDE auto-completion should be able to show the signature at any call site, and it would add some small overhead too.

Reply:
I see the attraction, but I think there are too many variants to make this worthwhile.  At least for these cases I'd guess we're pretty safe, in that any confusion of the roles of (here) `heap`, `dst`, `src` and `len` is highly likely to cause the generated code to crash during testing.

Reply:
I'm a bit torn here as well, but I think I agree that adding an argument struct for each of these methods would be too much overhead and the likelyhood of an issue escaping tests is minimal.